ddl/worker.go

```diff
@@ -245,29 +284,46 @@ func (d *ddl) runJob(job *model.Job) error {
 	var err error
 	switch job.Type {
 	case model.ActionCreateSchema:
+		err = d.onCreateSchema(t, job)
 	case model.ActionDropSchema:
+		err = d.onDropSchema(t, job)
 	case model.ActionCreateTable:
+		err = d.onCreateTable(t, job)
 	case model.ActionDropTable:
+		err = d.onDropTable(t, job)
 	case model.ActionAddColumn:
+		err = d.onAddColumn(t, job)
 	case model.ActionDropColumn:
+		err = d.onDropColumn(t, job)
 	case model.ActionAddIndex:
+		err = d.onCreateIndex(t, job)
 	case model.ActionDropIndex:
-	case model.ActionAddConstraint:
-	case model.ActionDropConstraint:
+		err = d.onDropIndex(t, job)
 	default:
-		return errors.Errorf("invalid job %v", job)
+		// invalid job, cancel it.
+		job.State = model.JobCancelled
+		err = errors.Errorf("invalid job %v", job)
 	}
 
-	// if err and inner doesn't cancel job, return err.
+	// saves error in job, so that others can know error happens.
 	if err != nil {
+		// if job is not cancelled, we should log this error.
 		if job.State != model.JobCancelled {
-			return errors.Trace(err)
+			log.Errorf("run job err %v", errors.ErrorStack(err))
 		}
 
 		job.Error = err.Error()
-	} else {
-		job.State = model.JobDone
+		job.ErrorCount++
 	}
+}
 
-	return nil
+// for every lease seconds, we will re-update the whole schema, so we will wait 2 * lease time
+// to guarantee that all servers have already updated schema.
+func (d *ddl) waitSchemaChanged() {
+	if d.lease > 0 {
```

Comment:
for local storage, lease is 0. 
SetSchemaLease will be used later in tidb server, so that we can control the lease in test with hbase storage later.